src/prefect/environments/storage/s3.py

```diff
@@ -0,0 +1,154 @@
+import io
+from typing import TYPE_CHECKING, Any, Dict
+
+
+import cloudpickle
+import pendulum
+from slugify import slugify
+
+import prefect
+from prefect.environments.storage import Storage
+
+if TYPE_CHECKING:
+    from prefect.core.flow import Flow
+
+
+class S3(Storage):
+    """
+    S3 storage class.  This class represents the Storage interface for Flows stored as
+    bytes in an S3 bucket.
+
+    This storage class optionally takes a `key` which will be the name of the Flow object
+    when stored in S3. If this key is not provided the Flow upload name will take the form
+    `slugified-flow-name/slugified-current-timestamp`.
+
+    Args:
+        - aws_access_key_id (str, optional): AWS access key id for connecting to S3.
+            Defaults to the value set in the environment variable
+            `AWS_ACCESS_KEY_ID` or `None`
+        - aws_secret_access_key (str, optional): AWS secret access key for connecting to S3.
+            Defaults to the value set in the environment variable
+            `AWS_SECRET_ACCESS_KEY` or `None`
+        - aws_session_token (str, optional): AWS session key for connecting to S3
```

Comment:
As we discussed offline, I'm still on the fence about the best auth pattern here.  When building locally, anything can go, but whenever this Storage needs to extract a flow object while running in Cloud, it's much tougher to specify things like env vars in that environment

Reply:
Yeah I need to think about this some more 🤔